OpenRA.Mods.Cnc/Traits/TransformsNearResources.cs

```diff
@@ -0,0 +1,128 @@
+
+namespace OpenRA.Mods.Cnc.Traits
+{
+	[Desc("Replace with another actor when a resource spawns adjacent.")]
+	public class TransformsNearResourcesInfo : ITraitInfo
+	{
+		[FieldLoader.Require]
+		[ActorReference]
+		public readonly string IntoActor = null;
+
+		public readonly CVec Offset = CVec.Zero;
+
+		[Desc("Don't render the make animation.")]
+		public readonly bool SkipMakeAnims = false;
+
+		[FieldLoader.Require]
+		[Desc("Resource type which triggers the transformation.")]
+		public readonly string Type = null;
+
+		[Desc("Resource density threshold which is required.")]
+		public readonly int Density = 1;
+
+		[Desc("This many adjacent resource tiles are required.")]
+		public readonly int Adjacency = 1;
+
+		[Desc("The number of ticks between each check.")]
+		public readonly int NumberOfTicksBetween = 25;
+
+		[Desc("The median time (in ticks) until the transformation starts.")]
+		public readonly int MedianTicks = 3000;
+
+		public virtual object Create(ActorInitializer init) { return new TransformsNearResources(init.Self, this); }
+	}
+
+	public class TransformsNearResources : ITransformsNearResources, INotifyAddedToWorld, INotifyRemovedFromWorld
+	{
+		readonly TransformsNearResourcesInfo info;
+		readonly ResourceLayer resourceLayer;
+		readonly Actor self;
+		readonly World world;
+		readonly double probability;
+		int ticks;
+
+		public TransformsNearResources(Actor self, TransformsNearResourcesInfo info)
+		{
+			this.info = info;
+			this.self = self;
+			world = self.World;
+			resourceLayer = world.WorldActor.Trait<ResourceLayer>();
+			probability = (1 - Math.Pow(2, (double)-info.NumberOfTicksBetween / info.MedianTicks)) * 10000;
+		}
+
+		void INotifyAddedToWorld.AddedToWorld(Actor self)
+		{
+			resourceLayer.AddTransformable(this);
+		}
+
+		void INotifyRemovedFromWorld.RemovedFromWorld(Actor self)
+		{
+			resourceLayer.RemoveTransformable(this);
+		}
+
+		public bool Transform()
+		{
+			if (++ticks <= info.NumberOfTicksBetween)
+				return false;
+
+			ticks = 0;
+
+			var adjacent = 0;
+			foreach (var direction in CVec.Directions)
+			{
+				var location = self.Location + direction;
+
+				var resource = resourceLayer.GetResourceType(location);
+				if (resource == null || resource.Info.Type != info.Type)
+					continue;
+
+				var density = resourceLayer.GetResourceDensity(location);
+				if (density < info.Density)
+					continue;
+
+				if (++adjacent >= info.Adjacency)
+					break;
+			}
+
+			if (adjacent < info.Adjacency)
+				return false;
+
+			var next = self.World.SharedRandom.Next(10000);
+			if (next > probability)
+				return false;
+
+			Transform(self);
+			return true;
+		}
+
+		void Transform(Actor self)
+		{
+			var transform = new Transform(self, info.IntoActor);
+
+			var facing = self.TraitOrDefault<IFacing>();
+			if (facing != null)
+				transform.Facing = facing.Facing;
+
+			transform.SkipMakeAnims = info.SkipMakeAnims;
+			transform.Offset = info.Offset;
+
+			self.CancelActivity();
```

Comment:
This should preferably be done by adding a `false` parameter to the `QueueActivity` call below instead.